### PR TITLE
C4 model core diagrams

### PR DESCRIPTION
**Pull request by eskatos**
Core [C4 Model](https://c4model.com/#coreDiagrams) diagrams for the Gradle Kotlin DSL. This should provide a good overview of systems at play and the Kotlin DSL architecture

The diagrams are Plant-UML (`.puml`) files using [C4-PlantUML](https://github.com/RicardoNiepel/C4-PlantUML).

A [README](https://github.com/gradle/kotlin-dsl/blob/eskatos/doc/c4/doc/c4/README.adoc) displays the rendered versions.

### CODE REVIEW

**Comment by JLLeitschuh** [comment]
The `C4_3_Component.png` is so low quality as to almost be unhelpful. Can you increase the quality?

Also, why are the edges of the boxes so weird? The sides aren't straight and they just end up being distracting.

c4_2_container

c4_2_container

Are the edges supposed to be wavy? If so, the pixel count is too low for it to look artful. Maybe it's just that the pixel count needs to be increased?

**Comment by eskatos** [comment]
@JLLeitschuh, moved to SVG files in the README and removed `LAYOUT_AS_SKETCH`
also added a tiny build that automates the svg rendering

**Comment by JLLeitschuh** [comment]
@eskatos Much, much better!

File: doc/c4/C4_3_Component.puml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+Rel(pluginPlugin, kgp, "Applies and configures")
+
+Rel(scriptResolver, tapiModels, "Requests", "TAPI")
+Rel(gradleKotlinDslApi, kotlinc, "Uses", "Embedded")
+
+Rel(ide, kotlinc, "Uses")
+
+Rel_L(kgp, buildCache, "Uses")
+Rel(providerExecution, buildCache, "Uses")
+Rel(pluginPlugin, buildCache, "Uses")
```
[the remark this answers is not in the document]
Comment (eskatos) [reply doc/c4/C4_3_Component.puml]: There are tasks in there that rely on the build cache. It felt worth mentioning at this level of detail.

[the remark this answers is not in the document]
Comment (JLLeitschuh) [reply doc/c4/C4_3_Component.puml]: Is there any way you can provide a more informative description of this? I had the same question but I didn't know how to ask it.

[the remark this answers is not in the document]
Comment (bamboo) [reply doc/c4/C4_3_Component.puml]: I think a reference to the build cache would make sense when talking about script compilation, here it seems to raise more questions than answers.

**Comment by eskatos** [comment]
Thanks for the reviews @JLLeitschuh @bamboo

**Pull request merged**